pydy/codegen/ode_function_generator.py

```diff
@@ -0,0 +1,705 @@
+        first order ordinary differential equations in one of two forms:
+
+        x' = f(x, t, p)
+
+        or
+
+        x' = f(x, t, r, p)
+
+        """
+        if self.system_type == 'full rhs':
+            self.generate_full_rhs_function()
+        elif self.system_type == 'full mass matrix':
+            self.generate_full_mass_matrix_function()
+        elif self.system_type == 'min mass matrix':
+            self.generate_min_mass_matrix_function()
+
+        return self.create_rhs_function()
+
+
+class CythonODEFunctionGenerator(ODEFunctionGenerator):
+
+    def __init__(self, *args, **kwargs):
+
+        if Cython is None:
+            raise Exception('Cython must be installed to use this class.')
+        else:
+            super(CythonODEFunctionGenerator, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def _cythonize(outputs, inputs):
+        return CythonMatrixGenerator(inputs, outputs).compile()
+
+    def _set_eval_array(self, f):
+
+        if self.specifieds is None:
+            self.eval_arrays = lambda q, u, p: f(q, u, p, *self.empties)
+        else:
+            self.eval_arrays = lambda q, u, r, p: f(q, u, r, p, *self.empties)
+
+    def generate_full_rhs_function(self):
+
+        self._define_inputs()
+        outputs = [self.right_hand_side]
+
+        self.empties = (np.empty(self.num_states, dtype=float),)
+
+        self._set_eval_array(self._cythonize(outputs, self.inputs))
+
+    def generate_full_mass_matrix_function(self):
+
+        self._define_inputs()
+        outputs = [self.mass_matrix, self.right_hand_side]
+
+        mass_matrix_result = np.empty(self.num_states ** 2, dtype=float)
+        rhs_result = np.empty(self.num_states, dtype=float)
+
+        self.empties = (mass_matrix_result, rhs_result)
+
+        self._set_eval_array(self._cythonize(outputs, self.inputs))
+
+    def generate_min_mass_matrix_function(self):
+
+        self._define_inputs()
+        outputs = [self.mass_matrix, self.right_hand_side,
+                   self.coordinate_derivatives]
+
+        mass_matrix_result = np.empty(self.num_speeds ** 2, dtype=float)
+        rhs_result = np.empty(self.num_speeds, dtype=float)
+        kin_diffs_result = np.empty(self.num_coordinates, dtype=float)
+        self.empties = (mass_matrix_result, rhs_result, kin_diffs_result)
+
+        self._set_eval_array(self._cythonize(outputs, self.inputs))
+
+
+class LambdifyODEFunctionGenerator(ODEFunctionGenerator):
+
+    def _lambdify(self, outputs):
+        # TODO : We could forgo this substitution for speed purposes and
+        # have lots of args for lambdify (like it used to be done) but there
+        # may be some limitations on number of args.
+        subs = {}
+        vec_inputs = []
+        if self.specifieds is None:
+            def_vecs = ['q', 'u', 'p']
+        else:
+            def_vecs = ['q', 'u', 'r', 'p']
+
+        for syms, vec_name in zip(self.inputs, def_vecs):
+            v = sm.DeferredVector(vec_name)
```

Comment:
Ok, thanks. If you do I'll support it, but I'll have to support both unless we want to up the SymPy dependency version for PyDy.